packages/descendants/README.md

```diff
@@ -0,0 +1,200 @@
+---
+last_updated: 2024-02-01
+---
+
+# Descendants
+
+![npm (scoped)](https://img.shields.io/npm/v/@leafygreen-ui/descendants.svg)
+
+#### [View on MongoDB.design](https://www.mongodb.design/component/descendants/example/)
+
+## Installation
+
+### Yarn
+
+```shell
+yarn add @leafygreen-ui/descendants
+```
+
+### NPM
+
+```shell
+npm install @leafygreen-ui/descendants
+```
+
+## Overview
+
+`Descendants` is an internal utility that allows components to track all rendered descendants.
+
+This is useful when developing menus in order to track when items are rendered/un-rendered.
+
+## Usage
+
+There are 4 steps required to set up a pair of components as Parent/Descendent.
+
+```tsx
+/**
+ * 1. Create a new Context
+ *
+ * We need to create a new Context object
+ * in order for the parent & child to know their relationships.
+ *
+ * Without this, a descendant won't know what parent context it belongs to.
+ * This also enables us to nest different descendant contexts
+ *
+ * e.g.
+ * A Menu that a second fly-out menu
+ * shouldn't track fly-out menu items as its own descendants)
+ *
+ */
+const MyDescendantsContext = createDescendantsContext('MyDescendantsContext');
+
+export const MyParent = ({ children, ...rest }: ComponentProps<'div'>) => {
+  /**
+   * 2. Initialize an empty descendants list and setter
+   *
+   * We call this _outside_ the Provider
+   * so we can access the `descendants` object
+   * from the Parent level.
+   */
+  const { descendants, dispatch } = useInitDescendants<HTMLDivElement>();
```

Comment:
what's the dispatch doing here? do we need to de-structure just to pass it back in? can we do that for the user automatically? when would a consumer want to use this directly?

Reply:
So that's the unfortunate part about this approach. If we design it so that you initialize the descendants object _outside_ the provider, you need a way to pass the dispatch function into the provider.

You could technically write something like this, but IMO it's not much better (and misleading, since `rest` only refers to one property)

```tsx
const { descendants, ...rest } = useInitDescendants<HTMLDivElement>();

<DescendantsProvider
  context={MyDescendantsContext}
  descendants={descendants}
  {...rest}
>

```